In [366]:
import numpy as np
import pandas as pd

In [409]:
df = pd.read_csv(r'C:\Users\User\Downloads\hw_work.csv')
df

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4             99.0             16.78   
1                27.47              195.5            103.0             16.62   
2                41.38              121.2            110.0             10.30   
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5            126.0             18.32   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

In [368]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3323 non-null   float64
 7   Total day calls         3323 non-null   float64
 8   Total day charge        3315 non-null   float64
 9   Total eve minutes       3324 non-null   float64
 10  Total eve calls         3325 non-null   float64
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3332 non-null   float64
 14  Total night charge      3333 non-null   

## RangeIndex: 3333 entries, 0 to 3332
## Data columns (total 20 columns):
## memory usage: 520.9+ KB
## כל מה שהטוטאל שלו בעמודת NON-NULL הוא פחות מ3333- בהכרח יש בו רכיב באחת השורות שחסר ערך. לדוגמה: 
## 9   Total eve minutes       3324 non-null   float64

In [370]:
df['Account length'].describe()

count    3333.000000
mean      101.064806
std        39.822106
min         1.000000
25%        74.000000
50%       101.000000
75%       127.000000
max       243.000000
Name: Account length, dtype: float64

In [371]:
df.describe().transpose()

count        mean        std     min     25%     50%  \
Account length          3333.0  101.064806  39.822106    1.00   74.00  101.00   
Area code               3333.0  437.182418  42.371290  408.00  408.00  415.00   
Number vmail messages   3333.0    8.099010  13.688365    0.00    0.00    0.00   
Total day minutes       3323.0  179.787150  54.419625    0.00  143.70  179.40   
Total day calls         3323.0  100.456214  20.057356    0.00   87.00  101.00   
Total day charge        3315.0   30.557831   9.255987    0.00   24.42   30.50   
Total eve minutes       3324.0  201.033935  50.676652    0.00  166.60  201.40   
Total eve calls         3325.0  100.110677  19.932115    0.00   87.00  100.00   
Total eve charge        3333.0   17.083540   4.310668    0.00   14.16   17.12   
Total night minutes     3333.0  200.872037  50.573847   23.20  167.00  201.20   
Total night calls       3332.0  100.115246  19.566710   33.00   87.00  100.00   
Total night charge      3333.0    9.039325   2.275873    1.04    7.52    9.05   
Total intl minutes      3333.0   10.237294   2.791840    0.00    8.50   10.30   
Total intl calls        3333.0    4.479448   2.461214    0.00    3.00    4.00   
Total intl charge       3328.0    2.764588   0.754086    0.00    2.30    2.78   
Customer service calls  3328.0    1.563101   1.315870    0.00    1.00    1.00   

                            75%     max  
Account length          127.000  243.00  
Area code               510.000  510.00  
Number vmail messages    20.000   51.00  
Total day minutes       216.500  350.80  
Total day calls         114.000  165.00  
Total day charge         36.780   59.64  
Total eve minutes       235.325  363.70  
Total eve calls         114.000  170.00  
Total eve charge         20.000   30.91  
Total night minutes     235.300  395.00  
Total night calls       113.000  175.00  
Total night charge       10.590   17.77  
Total intl minutes       12.100   20.00  
Total intl calls          6.000   20.00  
Total intl charge         3.270    5.40  
Customer service calls    2.000    9.00

In [372]:
df.head(5)

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     26              161.6            123.0   
2                      0              243.4            114.0   
3                      0              299.4             71.0   
4                      0              166.7            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             27.47              195.5            103.0             16.62   
2             41.38              121.2            110.0             10.30   
3             50.90               61.9             88.0              5.26   
4             28.34              148.3            122.0             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7               91.0               11.01   
1                254.4              103.0               11.45   
2                162.6              104.0                7.32   
3                196.9               89.0                8.86   
4                186.9              121.0                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                     1.0  False  
1                     1.0  False  
2                     0.0  False  
3                     2.0  False  
4                     3.0  False

In [373]:
df.tail(5)

State  Account length  Area code International plan Voice mail plan  \
3328    AZ             192        415                 No             Yes   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3328                     36              156.2             77.0   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3328             26.55              215.5            126.0             18.32   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
3328                279.1               83.0               12.56   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3328                 9.9                 6               2.67   
3329                 9.6                 4               2.59   
3330                14.1                 6               3.81   
3331                 5.0                10               1.35   
3332                13.7                 4               3.70   

      Customer service calls  Churn  
3328                     2.0  False  
3329                     3.0  False  
3330                     2.0  False  
3331                     2.0  False  
3332                     0.0  False

In [389]:
df[['State', 'International plan']]
df[df['International plan'] == 'Yes'].index
df[(df['International plan'] == 'Yes') & (df['Churn'] == True)].index

Index([  41,  115,  144,  198,  214,  230,  241,  258,  319,  354,
       ...
       3157, 3189, 3209, 3241, 3243, 3246, 3255, 3291, 3304, 3320],
      dtype='int64', length=137)

In [423]:
# df.iloc[14]
# df.iloc[14:]

# df.set_index('State', inplace=True)
# df.loc['LA']
# df.reset_index()

# df.drop('Area code', axis = 1, inplace=True)
# # df
# df.drop(df.index[3328], axis=0, inplace=True)
# df

# df['Total_min'] = df['Total intl minutes'].add(df['Total night minutes'], fill_value=0)
# df['mean(avg)'] = df['Total_min']/2
# df

df_head = df.head(1)
df_tail = df.tail(1)
concat_df = pd.concat([df_head, df_tail], axis=0, ignore_index=True)
concat_df

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    TN              74        415                 No             Yes   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1            110.0   
1                     25              234.4            113.0   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4             99.0             16.78   
1             39.85              265.9             82.0             22.60   

   Total night minutes  Total night calls  Total night charge  \
0                244.7               91.0               11.01   
1                241.4               77.0               10.86   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3                2.7   
1                13.7                 4                3.7   

   Customer service calls  Churn  
0                     1.0  False  
1                     0.0  False

In [439]:
dict = {'State': 'LA',
'Account length': 100,
'Area code': 416,
'International plan': 'Yes',
'Voice mail plan': 'Yes',
'Number vmail messages': 25,
'Total day minutes': 309.5,
'Total day calls': 117,
'Total day charge': 45,
'Total eve minutes': 209,
'Total eve calls': 101,
'Total eve charge': 194,
'Total night minutes': 233,
'Total night calls':322,
'Total night charge':5432,
'Total intl minutes':43,
'Total intl calls':22,
'Total intl charge':222,
'Customer service calls':2,
'Churn': False}


dict_pd = pd.DataFrame([dict])
pd_concat = pd.concat([df, dict_pd])
pd_concat

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
...    ...             ...        ...                ...             ...   
3329    WV              68        415                 No              No   
3330    RI              28        510                 No              No   
3331    CT             184        510                Yes              No   
3332    TN              74        415                 No             Yes   
0       LA             100        416                Yes             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
0                        25              265.1            110.0   
1                        26              161.6            123.0   
2                         0              243.4            114.0   
3                         0              299.4             71.0   
4                         0              166.7            113.0   
...                     ...                ...              ...   
3329                      0              231.1             57.0   
3330                      0              180.8            109.0   
3331                      0              213.8            105.0   
3332                     25              234.4            113.0   
0                        25              309.5            117.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0                45.07              197.4             99.0             16.78   
1                27.47              195.5            103.0             16.62   
2                41.38              121.2            110.0             10.30   
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
...                ...                ...              ...               ...   
3329             39.29              153.4             55.0             13.04   
3330             30.74              288.8             58.0             24.55   
3331             36.35              159.6             84.0             13.57   
3332             39.85              265.9             82.0             22.60   
0                45.00              209.0            101.0            194.00   

      Total night minutes  Total night calls  Total night charge  \
0                   244.7               91.0               11.01   
1                   254.4              103.0               11.45   
2                   162.6              104.0                7.32   
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
...                   ...                ...                 ...   
3329                191.3              123.0                8.61   
3330                191.9               91.0                8.64   
3331                139.2              137.0                6.26   
3332                241.4               77.0               10.86   
0                   233.0              322.0             5432.00   

      Total intl minutes  Total intl calls  Total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3329                 9.6                 4            

### calling the inplace parameter is for changing the actual table/pd. for example:

In [456]:
table = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
table

# drop without inplace- changes but not the original
new_table = table.drop('A', axis=1)
table

# with inpace- changes the origin
new_table_2 = table.drop('A', axis=1, inplace=True)
table


B
0  4
1  5
2  6

## אינדקס הוא דמוי מפתח
כלומר: יכולים להיות כמה אינדקסים דומים- אך עמודת אינדקס יש רק אחת והיא אמצעי הזיהוי של כל שורה בעמודות האחרות. זה בעצם כמו העמודה הראשית.
זיהוי העמודה הראשית היא על פי הכותרת- ולכן היא לא יכולה להיות שווה בשמה לכותרות של עמודות אחרות

axis0- rows
axis1- columns

In [473]:
# & == and
df[(df['International plan'] == 'Yes') & (df['Churn'] == True)].index
# | == or
df[(df['International plan'] == 'Yes') | (df['Churn'] == True)].index
# ~ == not
df[~(df['International plan'] == 'Yes')].index
# isin
df[df['State'].isin(['LA', 'WY'])]
# between
df[df['Account length'].between(50, 100)]


State  Account length  Area code International plan Voice mail plan  \
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
10      IN              65        415                 No              No   
11      RI              74        415                 No              No   
13      MT              95        510                 No              No   
...    ...             ...        ...                ...             ...   
3325    OH              78        408                 No              No   
3326    OH              96        415                 No              No   
3327    SC              79        415                 No              No   
3329    WV              68        415                 No              No   
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3                         0              299.4             71.0   
4                         0              166.7            113.0   
10                        0              129.1            137.0   
11                        0              187.7            127.0   
13                        0                NaN              NaN   
...                     ...                ...              ...   
3325                      0              193.4             99.0   
3326                      0              106.6            128.0   
3327                      0              134.7             98.0   
3329                      0              231.1             57.0   
3332                     25              234.4            113.0   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3                50.90               61.9             88.0              5.26   
4                28.34              148.3            122.0             12.61   
10               21.95              228.5             83.0             19.42   
11               31.91              163.4            148.0             13.89   
13                 NaN              247.6             75.0             21.05   
...                ...                ...              ...               ...   
3325             32.88              116.9             88.0              9.94   
3326             18.12              284.8             87.0             24.21   
3327             22.90              189.7             68.0             16.12   
3329             39.29              153.4             55.0             13.04   
3332             39.85              265.9             82.0             22.60   

      Total night minutes  Total night calls  Total night charge  \
3                   196.9               89.0                8.86   
4                   186.9              121.0                8.41   
10                  208.8              111.0                9.40   
11                  196.0               94.0                8.82   
13                  192.3              115.0                8.65   
...                   ...                ...                 ...   
3325                243.3              109.0               10.95   
3326                178.9               92.0                8.05   
3327                221.4              128.0                9.96   
3329                191.3              123.0                8.61   
3332                241.4               77.0               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
10                  12.7                 6               3.43   
11                   9.1                 5               2.46   
13                  12.3                 5               3.32   
...                  ...               ...                ...   
3325                 9.3                 4            